log/stat/stat.go

```diff
@@ -0,0 +1,107 @@
+package stat
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"text/tabwriter"
+
+	"github.com/peak/s5cmd/strutil"
+)
+
+const (
+	totalCount = iota
+	succCount
+)
+
+var (
+	enabled bool
+	stats   statistics
+)
+
+type statistics [2]syncMap
+
+// InitStat initializes collecting program statistics.
+func InitStat() {
+	enabled = true
+	for i := range stats {
+		stats[i] = syncMap{
+			Mutex:       sync.Mutex{},
+			mapStrInt64: map[string]int64{},
+		}
+	}
+}
+
+// syncMap is a statically typed and synchronized map.
+type syncMap struct {
+	sync.Mutex
+	mapStrInt64 map[string]int64
+}
+
+func (s *syncMap) add(key string, val int64) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.mapStrInt64[key] += val
+}
+
+// Stat is for storing a particular statistics.
+type Stat struct {
+	Visited    string `json:"visited"`
+	SuccVisits int64  `json:"success visits"`
+	ErrVisits  int64  `json:"error visits"`
+}
+
+// Collect collects function execution data.
+func Collect(path string, err *error) func() {
+	return func() {
+		if !enabled {
+			return
+		}
+		if err == nil || *err == nil {
+			stats[succCount].add(path, 1)
+		}
+		stats[totalCount].add(path, 1)
+	}
+}
+
+// Stats implements log.Message interface.
+type Stats []Stat
+
+func (s Stats) String() string {
+	b := bytes.Buffer{}
+
+	w := tabwriter.NewWriter(&b, 5, 0, 5, ' ', tabwriter.AlignRight)
+
+	fmt.Fprintf(w, "\n%s\t%s\t%s\t%s\t\n", "Operation", "Total", "Error", "Success")
+	fmt.Fprintf(w, "%s\t%s\t%s\t%s\t\n", "=========", "=====", "=====", "=======")
+	for _, stat := range s {
+		fmt.Fprintf(w, "%s\t%d\t%d\t%d\t\n", stat.Visited, stat.ErrVisits+stat.SuccVisits, stat.ErrVisits, stat.SuccVisits)
+	}
+
+	w.Flush()
+	return b.String()
+}
+
+func (s Stats) JSON() string {
+	return strutil.JSON(s)
```

Comment:
When I test --json flag, it prints following output:

```
[{"visited":"ls","success visits":2,"error visits":0}]
```

Can we use a more readable format like {"operation": "ls", "success": 2, "error": 0}? Also, it returns an array of jsons ( single json would be better).

Reply:
thanks. done.